accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,259 @@
+                                REFERENCES suseMinionInfo (server_id)
+                                ON DELETE SET NULL,
+    mac_address         VARCHAR(17),
+    root_device         VARCHAR(36),
+    salt_device         VARCHAR(36),
+    boot_image          VARCHAR(256),
+    kernel_parameters   VARCHAR(512)
+)
+```
+
+To prevent storing long `root_device` and `salt_device` we propose to **change saltboot use from dev-by-path to dev-by-uuid** and sending only UUIDs of used devices.
+
+
+`boot_image` points to part of built retail PXE OS image. Data about images are currently not available in database itself and are only available in generated pillar data. Information from pillar addressed by `boot_image` is then used in generated PXE entry to correctly set what initrd and kernel should be served to the terminal during PXE boot.
+
+`kernel_parameters` are taken from HWTYPE group given terminal is member of or from defaults passed on kernel command line. It is sent in event, because pxe server, where PXE entries are to be generated, does not have access to pillars from saltboot formula of HWTYPE group.
+
+`pxeserver_id` column references PXE server (usually branch server) which provides PXE/UEFI booting data (including UEFI HTTP boot if configured).
+
+Column `mac_address` specify MAC address of the terminal. Similar column is also present in `rhnServerNetInterface` table, however that information is populated only after hardware refresh is scheduled, which is done only after successful deployment. Whereas pxe entries needs to be available before full deployment because of cases when kernel version of inird is different to kernel version in image and terminal requires reboot to correct kernel version.
+
+Another option is that event handler creates the correct entry in `rhnServerNetInterface` table.
+
+### Detecting pxeserver_id
+
+In scenarios where terminals are connected to the branch server proxy, `pxeserver_id` is currently assumed to be branch server proxy and can be looked up in database by event handler. However architectures where proxy server is not involved such detection will not work (for example Hub scenarios where peripheral server is used instead of proxy server).
+
+We propose to solve this by **adapting saltboot to send FQDN of `next-server`**, that means server which provided initrd and kernel during PXE boot. This can be obtained by parsing DHCP response for `next-server` field. DHCP response is stored and available in dracut environment where saltboot is operating. Mandatory requirement in this case would be that **reverse lookup of IP address must be correctly configured** as `next-server` fields is IP address.
+
+Alternative is to use DNS service to lookup `tftp` CNAME, this is discussed in alternatives below.
+
+When the FQDN of `next-server` is not in the Uyuni database, such is the case when terminal is connected directly to the Uyuni, then `pxeserver_id` would be set to `null`.
```

Comment:
Terminal connected directly to Uyuni can be dectected more reliably - Uyuni server should know it's own FQDN.

Reply:
True, but the database entry is a reference to server id and AFAIK that would be populated only if uyuni is registered to itself. So it can be detected that fqdn is uyuni or something else but will result it same `pxeserver_id` entry.

Reply:
Where would we save such an FQDN, assuming `pxeserver_id` is `NULL`?

Reply:
> I am assuming FQDN of Uyuni server is generally known.

That assumption is correct, although the implementation is not ideal (we have a fragility problem there)